### DISCUSSION TITLE
Dynamic Hotspots

### DISCUSSION THREAD

**Discussion by visualgroup** [google/model-viewer#4116] [Q&A] [answered]
Could someone explain the data returned from surfaceFromPoint(clientX, clientY) for the newly implemented dynamic hotspots? eg data-surface="0 0 228 113 111 0.217 0.341 0.442".
What do the eight numbers returned represent?

I currently create the hotspot data directly from within blender using the previous data-position, data-normal, data-orbit and data-target fields as I prefer picking exact vertices on the model rather than using the model-viewer editor which is not as precise. I'd like to be able to do the same for the data-surface field as well (if possible) by adding to my python scripts.

I tried comparing a default cube with data-surface output from the editor but could only discern some of the data but not all or their relation to the model by editing the returned data and seeing the affect on the hotspot.

Hoping the creators can give me some advice, thanks.

**Comment by elalish** [comment, marked as answer]
Sure, there's a very brief description [here](https://modelviewer.dev/docs/#entrydocs-annotations-slots-hotspot). The first two are glTF-style indices. The three vertices don't technically have to be from the same triangle, but it's strongly recommended. Let me know if you need any more detail. Also, if you have anything to share about your Blender process, I bet others here would love to know. Feel free to make a show & tell.

**Comment by visualgroup** [comment]
Perfect, thanks.... I missed that info in the docs. "Barycentric coordinates" is a new one on me :-)

My Blender Python code is pretty horrible but works well for my workflow. I'll give dynamic hotspot generation a go and then see if I can polish the code a bit before sharing.

For those that are interested the basic process is.

1. Run a script to add two vertex groups to the model called "camera_hotspot" and "no_camera_hotspot". First vertex group is for hotspots that will have additional data-orbit and data-target info added, second will just return data-position and data-normal info.
2. Select the vertices that you want to make hotspots with/without camera data and add them to the respective vertex groups.
3. Then run a script to generate the hotspots. For camera_hotspots this adds an empty at each of the selected vertices and names it hotspot-xx etc. Creates cameras and positions them a set distance from each of the selected vertices along the normal. Also adds  empties and names them camera_target-xx etc. Then runs through the vertices in the no_camera_vertex group and just adds an empty at each vertex.
4. Then you can move the camera_target empties to get the desired camera views for those hotspots where the Blender viewport through that camera represents the model_viewer view. You can tweak this to get the perfect view. Also means you can place it in space i.e it doesn't need to be on the mesh.
5. Then, once you have all the camera views set up as you want you run a script that generates all the required model-viewer HTML for the hotspots, slot names, classes, data-position, data-normal, data-target, data-orbit etc. This generates a .txt file which you just paste into your .html and bobs your uncle. data-position is just each vertex co-ordinate, data-normal is each vertex normal, data-target is the co-ordinate location of the target empty and data-orbit is the co-ordinate location of each camera that Blender can easily spit out but formatting it as full html hotspots via python is just icing on the cake.

Thanks again.

**Reply by LeClam** [reply]
Hey, I would be extremely interested in this Python script. Could you share this with me? I've been looking for a solution to set the hotspots directly in Blender for a while now.

Thanks a lot for your help
